### ISSUE TITLE
Possibility to change class visibility of generated files

### ISSUE THREAD

**Issue by FlowMo7** [google/protobuf-gradle-plugin#401]
Is there a possibility to change the generated classes to not be `public`, but e.g. `package private`?

I've stumbled across this in the following use-case:

I am developing a java library, which internally uses protocol buffers as serialization tool for data communication.
The classes generated from the `.proto` files should not be visible outside of the library, as they are not needed to be known / visible to the library consumer.

Is there a way to prevent the files from being exposed as public API from this library?

(What I've tried so far and did not work is to generate the classes in a seperate gradle module, and include this module using `implementation` (instead of `api`), but as the library is published into a maven artifactory, and afaik there is no differentiation between those two mechanisms in maven, the classes are still visible then for the end-library-consumer).

**Comment by ejona86** [comment]
Since `public` vs package-private is concerning the generated code itself, this is not the appropriate repository. For example, if you are talking about Java generated code for messages, https://github.com/protocolbuffers/protobuf would be the appropriate repository.

`implementation` vs `api` does provide some benefit on Maven Central. `implementation` dependencies show up as `runtime` dependencies in the pom, so they wouldn't be used by Maven/Gradle when compiling. A user could still add an explicit `compile` dependency on the dependency, though.

What we've done in gRPC when we need to hide protobufs is put the generated code in an `internal` package (e.g., `io.grpc.xds.internal`) or to shade it to an internal package (there is some generated code that we consume from Maven Central that is unstable).

**Comment by voidzcy** [comment]
You could try the [Java Library Distribution Plugin](https://docs.gradle.org/current/userguide/java_library_distribution_plugin.html) although it is incubating.

**Comment by FlowMo7** [comment]
Many thanks for the reply!
(I was indeed unsure whether this repo or the protobuf repo were the right one to submit this issue :slightly_smiling_face: - I will consider raising this issue in the protobuf repo if this issue still exists for me after your suggestions)

I still have to check your solution about the runtime maven dependencies in the pom or the shading plugin, I'm unfortunately not as fit in the whole maven dependency system as I would like to be, and unfortunately some other tasks took over priority on my end right now.

Moving the classes in a package named internal only is not an option for me right now, as the generated class names are nearly identical as the POJO classes in the public API, which would lead to too much confusion.

If someone's stumbling across this in the meantime and is interested in a _super hacky_ way I've workarounded this so far: I've added a gradle task after the `generateProto` task which visits the generated classes and manually strips out the public modifier of the classes.
This combined with a [binary-compatibility-validator](https://github.com/Kotlin/binary-compatibility-validator) (in which changes in the public API get reported automatically, in case the string-replace does not work for any reason) do the trick for me for now.
(Again, this is not a solution I would recommend nor aim for, but a quick and dirty workaround so far).
